akka-stream/src/main/scala/akka/stream/impl/FanIn.scala

```diff
@@ -183,13 +204,19 @@ private[akka] object FanIn {
     }
 
     def inputsAvailableFor(id: Int) = new TransferState {
-      override def isCompleted: Boolean = depleted(id) || cancelled(id)
-      override def isReady: Boolean = pending(id)
+      override def isCompleted: Boolean = {
+        val state = states(id)
+        has(state, Depleted) || has(state, Cancelled)
+      }
```

Comment:
I can definitely revert so that every flag check reads off the array, but I'm not sure how less brittle that is.